File: app/src/main/java/com/martinb/marvelapp/ui/MainActivity.java
package com.martinb.marvelapp.ui;

import android.Manifest;
import android.content.pm.PackageManager;
import android.net.Uri;
import android.support.v4.app.ActivityCompat;
import android.support.v4.app.DialogFragment;
import android.support.v4.app.FragmentManager;
import android.support.v4.app.FragmentTransaction;
import android.support.v4.content.ContextCompat;
import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;

import com.martinb.marvelapp.R;
import com.martinb.marvelapp.data.MarvelApiClient;
import com.martinb.marvelapp.data.model.Character;
import com.martinb.marvelapp.data.model.Result;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class MainActivity extends AppCompatActivity implements OnItemClicked,CharacterFragment.OnFragmentInteractionListener {

    private CharacterAdapter adapter;
    private RecyclerView recyclerView;
    private CharacterFragment characterDescription;
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        checkPermissionStatus();

        // set up the RecyclerView
        recyclerView = findViewById(R.id.baseRecycler);
        recyclerView.setLayoutManager(new LinearLayoutManager(this));

        MarvelApiClient.getInstance(getApplicationContext()).getData().enqueue(new Callback<Character>() {
            @Override
            public void onResponse(Call<Character> call, Response<Character> response) {
                Character character = response.body();
                setAdapter(character);
            }

            @Override
            public void onFailure(Call<Character> call, Throwable t) {
                t.printStackTrace();
            }
        });
    }

    private void setAdapter(Character character){
        adapter = new CharacterAdapter(this, character.getData().getResults());
        recyclerView.setAdapter(adapter);
        adapter.setOnItemClickedListener(this);
    }
    public void checkPermissionStatus() {
        if (ContextCompat.checkSelfPermission(this,
                Manifest.permission.INTERNET)
                != PackageManager.PERMISSION_GRANTED) {

            ActivityCompat.requestPermissions(this,
                    new String[]{Manifest.permission.INTERNET},
                    214);
        }
    }

    @Override
    public void onItemSelected(Result result) {
        characterDescription = CharacterFragment.newInstance("","");
        characterDescription.setSelectedCharacter(result);
        FragmentManager fm = getSupportFragmentManager();
        characterDescription.show(fm,"show");

    }

    @Override
    public void onFragmentInteraction(Uri uri) {

    }
}

File: app/src/main/java/com/martinb/marvelapp/ui/MainActivity.kt
package com.martinb.marvelapp.ui

import android.Manifest
import android.content.pm.PackageManager
import android.os.Bundle
import android.widget.Toast
import android.widget.Toast.LENGTH_SHORT
import androidx.appcompat.app.AppCompatActivity
import androidx.core.app.ActivityCompat
import androidx.core.content.ContextCompat
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout
import com.martinb.marvelapp.R
import com.martinb.marvelapp.data.model.Character
import com.martinb.marvelapp.data.model.Result
import com.martinb.marvelapp.di.marvelAppModule
import kotlinx.android.synthetic.main.activity_main.*
import org.koin.android.ext.android.inject
import org.koin.android.ext.android.startKoin

class MainActivity : AppCompatActivity(), MainView,OnItemClicked {

    private lateinit var adapter: CharacterAdapter
    private lateinit var characterDescriptionFragment : CharacterDescriptionFragment
    private lateinit var swipeRefreshLayout: SwipeRefreshLayout
    private val mainPresenter : MainPresenter by inject()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        startKoin(applicationContext, listOf(marvelAppModule))
        checkPermissionStatus()

        swipeRefreshLayout = findViewById(R.id.swiperefresh)

        baseRecycler.layoutManager = LinearLayoutManager(this)

        submit_search.setOnClickListener {
            if(!character_input.text.toString().equals(""))
                mainPresenter.getCharacterFilteredInfo(character_input.text.toString())
            else
                Toast.makeText(applicationContext,"Please, insert something", LENGTH_SHORT).show()
        }

        mainPresenter.attachView(this)
        swipeRefreshLayout.isRefreshing = true
        mainPresenter.getCharacterInfo()
    }


    private fun setAdapter(character: Character) {
        adapter = CharacterAdapter(this, character.data.results,this)
        baseRecycler.adapter = adapter
    }

    private fun checkPermissionStatus() {
        if (ContextCompat.checkSelfPermission(this,
                        Manifest.permission.INTERNET) != PackageManager.PERMISSION_GRANTED) {

            ActivityCompat.requestPermissions(this,
                    arrayOf(Manifest.permission.INTERNET),
                    214)
        }
    }

    override fun charactersInfo(character: Character?) {
        swipeRefreshLayout.isRefreshing = false
        character?.let { setAdapter(character)}
    }

    override fun onItemClicked(character: Result?) {
        character?.let {characterDescriptionFragment = CharacterDescriptionFragment.newInstance(character)}
        val fm = this@MainActivity.supportFragmentManager
        characterDescriptionFragment.show(fm,"description")
    }

    override fun filteredCharactersInfo(character: Character?) {
       character?.let { adapter.setNewCharactersData(character.data.results) }
   }

}
